Add clear_cofactor proof

bboston7 opened the pull request:
No rush here.  I'll take a look at the review when I'm back at work on 5/3!

bboston7 commented on proof/clear_cofactor_p2.saw:
```diff
@@ -0,0 +1,190 @@
+  {{ \x y -> neg E' (add E' x y) == add E' (neg E' x) (neg E' y) }};
+
+add_normalize_p2_thm <- admit_cryptol
+  {{ \P1 P2 -> add_E' (P1.0, Fp_2.normalize P1.1) P2 == add_E' P1 P2 }};
```
I based this theorem on this theorem from the `miller-loop` branch: https://github.com/GaloisInc/BLST-Verification/blob/90848304f3775de04d065d1dfbcc6a20b0920fbb/proof/algebraic-laws.saw#L85

I'm specifically calling it out because I'd like to double check that the same principle applies here too.

msaaltink replied:
I do not think this is true.  If P2 == point_O, then `add E' P1 p2` equals `P1`, so if P1 is not normalized this rule is incorrect.

bboston7 replied:
Is there something similar to `POINTonE2_invariant` and `POINTonE2_affine_invariant` for points of type `ProjectivePoint t_Fp_2` and `AffinePoint t_Fp_2` respectively?  In [this commit](https://github.com/GaloisInc/BLST-Verification/pull/51/commits/ff5e94ab7cad2a39424ff00024698e718742049d) I removed `add_normalize_p2_thm` and conditioned the proof on an undefined `ProjectivePoint_Fp_2_invariant`.  I also made use of an undefined `POINTonE2_affine_invariant` for some theorems.  I initially tried conditioning the proof on `POINTonE2_invariant (POINTonE2_rep P)` but this is always `True`.  The  important properties of these invariants would be:

1.  `forall P : AffinePoint_Fp_2_invariant P -> ((P.0, Fp_2.normalize P.1) == P)`
2. `forall P : AffinePoint_Fp_2_invariant P -> AffinePoint_Fp_2_invariant (-z * P)`
3. `forall P : AffinePoint_Fp_2_invariant (affinify_E' P) == ProjectivePoint_Fp_2_invariant P`

bboston7 replied:
> Is there something similar to POINTonE2_invariant and POINTonE2_affine_invariant for points of type ProjectivePoint t_Fp_2 and AffinePoint t_Fp_2 respectively?

I'm not sure this is even the right equivalence to ask about, hence the further explanation.

msaaltink replied:
Property 3 might be the tricky one.  The most obvious invariant properties for these types would be to assert that all of the coordinates satisfy the field invariant, `fp2_invariant`, which ought to give properties 1 and 2, but then I think affinify would always give something satisfying the invariant.   I suppose if we throw in "on_curve" then 3 might hold as well.  So maybe that's it: use `is_point_affine` and `is_point_projective`.

msaaltink commented on proof/clear_cofactor_p2.saw:
```diff
@@ -0,0 +1,190 @@
+  };
+
+// Algebraic theorems
+sub_to_add_neg_p2_thm <- admit_cryptol
```
You could prove this one; that's just the definition of `sub`.

msaaltink commented on proof/clear_cofactor_p2.saw:
```diff
@@ -0,0 +1,190 @@
+
+  crucible_execute_func [out_ptr, inp_ptr];
+
+  crucible_points_to out_ptr (crucible_term {{ POINTonE2_rep (clear_cofactor_E2_impl (POINTonE2_abs inp)) }});
```
I usually prefer to put the more abstract function in the spec, here that would be ` clear_cofactor_bls12381_g2 ` I think.  The proof of the override is no harder; you just start by using the `equiv` theorem to replace the abstract function with the more implementation-oriented one, and you've got that theorem already.

nano-o replied:
That can only be done with `llvm_postcond` though. Is that what you mean?

Btw, we should switch to the `llvm_` prefix (`crucible_` is deprecated).

msaaltink replied:
Oops, yes you are right.  I hadn't looked at the equiv theorem statement closely enough.  So this has to stay as it is.  That's OK and we've done it this way elsewhere.

msaaltink reviewed (changes requested):
Looks good, except for that one assumption that needs fixing.  Thumbs up for adding some tests for your Cryptol definitions.

bboston7 commented:
I removed the incorrect `add_normalize_p2_thm`, proved `sub_to_add_neg_p2_thm`, and replaced the `crucible` prefix in all of the commands I used with `llvm`.  @nano-o this branch is now ready for you to merge into your own branch.

nano-o commented on proof/clear_cofactor_p2.saw:
```diff
@@ -0,0 +1,210 @@
+  {{ \P -> if AffinePoint_Fp_2_invariant (affinify E' P)
+           then clear_cofactor_bls12381_g2 (affinify E' P) ==
+                affinify E' (clear_cofactor_E2_impl P)
+           else True }}
```
To use it as a rewrite rule, it would be easier to have
```
  {{ \P -> clear_cofactor_bls12381_g2 (affinify E' P) ==
      if AffinePoint_Fp_2_invariant (affinify E' P)
      then affinify E' (clear_cofactor_E2_impl P)
      else apply clear_cofactor_bls12381_g2 (affinify E' P) }}
```

bboston7 replied:
The commit I just pushed up changes the theorem to this form.

nano-o commented on proof/clear_cofactor_p2.saw:
```diff
@@ -0,0 +1,210 @@
+  {{ \P -> if AffinePoint_Fp_2_invariant (affinify E' P)
+           then clear_cofactor_bls12381_g2 (affinify E' P) ==
+                affinify E' (clear_cofactor_E2_impl P)
+           else True }}
```
Is the invariant assumption necessary or just convenient for the proof? One issue is that I know have to establish it at the call site.

bboston7 replied:
I just double checked and I believe this is necessary.  There are two double negatives on one side of the equality, and zero on the other.  So the goal will end up with `normalize` calls on only one side that need to be no-ops.
